portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -37,26 +40,45 @@
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
 
-  static String dataType = "Comment";
-  static String commentTimestamp = "timestamp";
-  static String commentTitle = "title";
-  static String maxCommentsParam = "max";
+  static String DATA_TYPE = "Comment";
+  static String COMMENT_TIMESTAMP = "timestamp";
+  static String COMMENT_AUTHOR = "name";
+  static String COMMENT_CONTENT = "message";
+  static String MAX_COMMENTS_PARAM = "max";
+  static String PAGE_NUM_PARAM = "page";
+  static int PAGE_SIZE;
+  static int PAGE_NUM;
 
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    int numComments = getNumComments(request);
-    Query query = new Query(dataType).addSort(commentTimestamp, SortDirection.DESCENDING);
+    PAGE_SIZE = getNumComments(request);
+    PAGE_NUM = Integer.parseInt(request.getParameter(PAGE_NUM_PARAM));
+
+    FetchOptions fetchOptions = FetchOptions.Builder.withLimit(PAGE_SIZE*PAGE_NUM);
```

Comment:
Please use .withOffset (http://cloud/appengine/docs/standard/java/javadoc/com/google/appengine/api/datastore/FetchOptions.Builder) rather than querying all results and performing calculations involving `numResults`. This change will make the code more understandable.

Reply:
Good job!